usb: disable previously enabled endpoint before re-enabling

mparanjpe opened the pull request:
When switching between alternate settings of an interface, it is currently possible to call `set_endpoint()` multiple times on an endpoint without first calling `reset_endpoint()`. For these situations, it is beneficial to track endpoints for which `set_endpoint()` has previously been called, and then reset them to properly terminate any transfers, and to return the HAL to the correct state

mparanjpe committed:
usb: disable previously enabled endpoint before re-enabling

When switching between alternate settings of an interface, it is
currently possible to call set_endpoint() multiple times on an endpoint
without first calling reset_endpoint(). For these situations, it is
beneficial to track endpoints for which set_endpoint() has previously
been called, and then reset them to properly terminate any transfers,
and to return the HAL to the correct state

Signed-off-by: Milind Paranjpe <mparanjpe@yahoo.com>

jfischer-no reviewed (changes requested):
We are working on the new USB support, device part should come into the tree in the next weeks. I would not like to change anything in this "legacy" code that is not bug-fix. What bug in the tree does it fix?

desowin commented on subsys/usb/device/usb_device.c:
```diff
@@ -530,6 +570,14 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
+	if (ep_mask & usb_dev.ep_set_mask) {
+		reset_endpoint(ep_desc);
+		/* allow any canceled transfers to terminate */
+		k_usleep(150);
```
`k_yield()` can return immediately and thus I believe it is not good way to enforce small delays.

mparanjpe replied:
@jfischer-no I did try `k_yield()` and it didn't work. That's why I kept the `k_usleep()`...

mparanjpe commented:
@jfischer-no 
> We are working on the new USB support, device part should come into the tree in the next weeks. I would not like to change anything in this "legacy" code that is not bug-fix. What bug in the tree does it fix?

Understood - if the underlying issue is fixed by the refactor, then I can close this PR.

My RT685 USB device application uses the `usb_transfer()` USB API to communicate with the USB host. After initialization is complete, a read transaction is queued to deal with any OUT traffic from the USB host. If the host decides to change to a different alternate setting, this pending transaction must be canceled. The current implementation of `usb_transfer()` does not appear to do enough to put the HAL in a state where subsequent read transactions can succeed. The only way I could make things work on my RT685 was to call `reset_endpoint()` in `usb_device.c`, then follow it with a `set_endpoint()` if that same endpoint had been enabled before.

mparanjpe commented on subsys/usb/device/usb_device.c:
```diff
@@ -135,6 +135,8 @@ static struct usb_dev_priv {
 	/** Remote wakeup feature status */
 	bool remote_wakeup;
+	/** Tracks whether set_endpoint() had been called on an EP */
+	uint32_t ep_set_mask;
```
OK let me change

mparanjpe commented on subsys/usb/device/usb_device.c:
```diff
@@ -508,6 +513,40 @@ static bool usb_get_descriptor(struct usb_setup_packet *setup,
+	/* check that endpoint index is in 0 to 15 range */
+	ep_idx = ep & (~USB_EP_DIR_IN);
+	if (ep_idx > 15) {
+		LOG_ERR("Endpoint 0x%02x is invalid", (int)ep);
```
I can remove - this is just me being pedantic for `%x`

mparanjpe commented on subsys/usb/device/usb_device.c:
```diff
@@ -522,6 +561,7 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 {
 	struct usb_dc_ep_cfg_data ep_cfg;
 	int ret;
+	uint32_t ep_mask;
```
Will reorder

mparanjpe commented on subsys/usb/device/usb_device.c:
```diff
@@ -569,6 +618,7 @@ static bool reset_endpoint(const struct usb_ep_descriptor *ep_desc)
 {
 	struct usb_dc_ep_cfg_data ep_cfg;
 	int ret;
+	uint32_t ep_mask;
```
Will reorder

mparanjpe commented on subsys/usb/device/usb_device.c:
```diff
@@ -1175,7 +1229,16 @@ static int foreach_ep(int (* endpoint_callback)(const struct usb_ep_cfg_data *))
 static int disable_interface_ep(const struct usb_ep_cfg_data *ep_data)
 {
-	return usb_dc_ep_disable(ep_data->ep_addr);
+	int ret;
+	uint32_t ep_mask;
```
Will reorder

jfischer-no commented:
> @jfischer-no
> 
> > We are working on the new USB support, device part should come into the tree in the next weeks. I would not like to change anything in this "legacy" code that is not bug-fix. What bug in the tree does it fix?
> 
> Understood - if the underlying issue is fixed by the refactor, then I can close this PR.

It is not a refactor, it is complete new implementation. Please continue on this PR.

> My RT685 USB device application uses the `usb_transfer()` USB API to communicate with the USB host. After initialization is complete, a read transaction is queued to deal with any OUT traffic from the USB host. If the host decides to change to a different alternate setting, this pending transaction must be canceled. The current implementation of `usb_transfer()` does not appear to do enough to put the HAL in a state where subsequent read transactions can succeed. The only way I could make things work on my RT685 was to call `reset_endpoint()` in `usb_device.c`, then follow it with a `set_endpoint()` if that same endpoint had been enabled before.

mparanjpe commented:
@jfischer-no 
> It is not a refactor, it is complete new implementation. Please continue on this PR.

I will get back to my PRs shortly - I have been busy work other work...

desowin commented:
Keeping track of enabled endpoints is not only needed for switching between alternate settings, but also for `usb_disable()`. That is, the `usb_disable()` should disable all endpoints that were enabled. This PR is really a part of bug fix for #52339. Disabling the endpoints alone won't fully fix the problem on nrfx, but is really necessary for correct solution. I'll base my fix to #52339 on top of this PR.

mparanjpe commented:
@desowin 
> Keeping track of enabled endpoints is not only needed for switching between alternate settings, but also for `usb_disable()`. That is, the `usb_disable()` should disable all endpoints that were enabled. This PR is really a part of bug fix for #52339. Disabling the endpoints alone won't fully fix the problem on nrfx, but is really necessary for correct solution. I'll base my fix to #52339 on top of this PR.

I have addressed the feedback - thanks for your patience...

jfischer-no commented on subsys/usb/device/usb_device.c:
```diff
@@ -146,6 +148,9 @@ struct usb_setup_packet_packed {
 	uint16_t wLength;
 } __packed;
 
+/* forward declaration of static functions */
```
Not addressed, please do not comment on obvious.

mparanjpe replied:
Removed

jfischer-no commented on subsys/usb/device/usb_device.c:
```diff
@@ -508,6 +513,38 @@ static bool usb_get_descriptor(struct usb_setup_packet *setup,
+ */
+static uint32_t get_ep_mask_from_addr(uint8_t ep)
+{
+	uint32_t ep_mask = 0;
```
I suggest to use ep_bm consistently.

mparanjpe replied:
Changed (I also changed the name of the function)

jfischer-no commented on subsys/usb/device/usb_device.c:
```diff
@@ -508,6 +513,38 @@ static bool usb_get_descriptor(struct usb_setup_packet *setup,
+		goto done;
+	}
+
+	ep_mask = 1 << ep_idx;
```
ep_mask = BIT(ep_idx);

you can remove the empty line below, or actually, I find this variant more readable:

```c
	if (ep & USB_EP_DIR_IN) {
		ep_bm = BIT(ep_idx + 16);
	} else {
		ep_bm = BIT(ep_idx);
	}
```

mparanjpe replied:
Thanks - changed
